Mathlib/Analysis/Convex/ConvexFunction.lean

```diff
@@ -0,0 +1,541 @@
+/-
+Copyright (c) 2023 Chenyi Li, Ziyu Wang, Zaiwen Wen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chenyi Li, Ziyu Wang, Penghao Yu, Zhipeng Cao, Shengyang Xu, Zaiwen Wen
+-/
+import Mathlib.Analysis.Calculus.FDeriv.Basic
+import Mathlib.Analysis.InnerProductSpace.Dual
+import Mathlib.Analysis.Convex.Function
+import Mathlib.Analysis.Calculus.MeanValue
+import Mathlib.Analysis.Calculus.Gradient.Basic
+
+/-!
+## Main results
+
+This file contains the following parts of smooth convex functions.
+* (First Order Condition for convex functions)
+    Function f(x) is convex if and only if for any x, y ∈ s, f(y) ≥ f(x) + f'(x) (y - x)
+* (Monotonity of Gradient)
+    Function f(x) is convex if and only if its gradient is monotone, i.e.
+    for any x, y ∈ s, (f'(x) - f'(y)) (x - y) ≥ 0
+* (Corresponding Result for strict convex functions)
+    Function f(x) is convex if and only if its gradient is strictly monotone, i.e.
+    for any x, y ∈ s, x ≠ y, (f'(x) - f'(y)) (x - y) ≥ 0
+
+Both results are stated in a gradient version and a fderiv version.
+-/
+
+variable {E : Type*} [NormedAddCommGroup E] [InnerProductSpace ℝ E] [CompleteSpace E]
+
+section FirstOrderCondition
+
+open InnerProductSpace
+
+variable {f : E → ℝ} {f' : E → (E →L[ℝ] ℝ)} {x y x': E} {s : Set E}
+
+private lemma point_proportion {a b: ℝ} (_ : 0 ≤ a) (_ : 0 ≤ b) (sumab: a + b = 1)
+  (hpoint : x' = a • x + b • y) :  x - x' =  b • (x - y) := by
+    calc
+      x - x' = x - (a • x + b • y):= by rw [hpoint]
+      _ = x - a • x - b • y:= sub_add_eq_sub_sub x (a • x) (b • y)
+      _ = (1 : ℝ) • x - a • x - b • y:= by rw[one_smul]
+      _ = (1 - a) • x - b • y:= by rw[sub_smul 1 a]
+      _ = b • x - b • y:= by rw [← sumab]; ring_nf
+      _ = b • (x - y):= Eq.symm (smul_sub b x y)
+
+theorem first_order_condition (h : HasFDerivAt f (f' x) x) (hf : ConvexOn ℝ s f)
+    (xs : x ∈ s) : ∀ (y : E), y ∈ s → f x + f' x (y - x) ≤ f y := by
```

Comment:
The separate spelling for `x` and `y` assumptions could be confusing. The following is much cleaner, especially when you already have a (y : E) declared above.
```suggestion
    (xs : x ∈ s) (hy : y ∈ s) : f x + f' x (y - x) ≤ f y := by
```